README
----------------------------------------------------------------------
    This is free software; you can redistribute it and/or modify it
    under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License , or
    (at your option) any later version.

    This software is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this software. If not, write to the Free Software
    Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
----------------------------------------------------------------------


1. DESCRIPTION
--------------

INetSim is a software suite for simulating common internet services
in a lab environment, e.g. for analyzing the network behaviour of
unknown malware samples.


1.1 Implemented service modules

Currently, modules for the simulation of the following services
are included with the INetSim distribution:

- HTTP/HTTPS
  - "real-mode": Delivers existing files from a webroot directory.
  - "fake-mode": Delivers configured fake files based on the file extension
                 in the HTTP request (e.g. .html or .exe) or static paths,
                 Requests for checkip.dyndns.org are answered with
                 client's IP address
  - supports HTTP methods GET, HEAD, POST and OPTIONS with HTTP/1.0 and HTTP/1.1
- SMTP/SMTPS
  - received e-mails are stored in mbox format
  - supports ESMTP and flexible configuration of service extensions
  - supports authentication methods PLAIN, LOGIN, ANONYMOUS, CRAM-MD5 and CRAM-SHA1
  - arbitrary authentication data is accepted and logged in plain text
- POP3/POP3S
  - dynamic creation of mailbox content from supplied mbox files
  - supports authentication methods PLAIN, LOGIN and CRAM-MD5
  - arbitrary authentication data is accepted and logged in plain text
- DNS
  - forward and reverse lookup with default and static configuration
- FTP/FTPS
  - download and upload
  - builds a virtual filesystem based on an existing ftproot directory
    which allows for creation and deletion of arbitrary files
- TFTP
  - download and upload
- IRC
  - basic command set
- NTP
- Ident
- Finger
- Syslog
- "Small servers":
  - Daytime,
  - Time,
  - Echo,
  - Chargen,
  - Discard and
  - Quotd
- Dummy

The listening port number can be configured for each service.


1.2 Faketime

INetSim can be run in 'faketime' mode to analyze the runtime behaviour
of malware which use NTP or Time/Daytime to start specific actions based
on the current date and time.

In 'faketime' mode, all services using date/time information (e.g. NTP or
HTTP) respond with a fake timestamp which is based on a configured delta
to current system time. Optionally, this delta can automatically be
incremented or decremented by a configured value at specific intervals.


1.3 Connection redirection

In addition to connection redirection via fake DNS responses, INetSim allows
for IP-based redirection of arbitrary connections (tcp, udp and icmp).
This feature is only available when running INetSim on Linux platforms
with Kernel support for packet queueing (Kernel compile time option
CONFIG_NETFILTER_NETLINK_QUEUE).

This feature supports static rules for connection redirection based on
target IP address, port and/or protocol. INetSim can also act as
NAT router for redirection of packets to other hosts. Optionally,
the TTL value of IP packets sent to the clients from different "virtual"
connection targets can be varied to make traffic look more authentic.


1.4 Dummy service

The Dummy service simply logs all data received from the client.
This module is most useful when used along with connection redirection
to capture all traffic sent from the client to ports not bound to any
other service module.

Optionally, a configurable banner string can be sent if no data
has been received for a given amount of time after the client
established the connection. This might be useful e.g. while analyzing
a malware which expects a POP3 or SMTP server on an unusual port.


1.5 Logging and reports

All incoming requests to the simulated services and the corresponding
outgoing replies are logged in detail. When stopping an INetSim session,
optionally an additional report for that session with a summary of the
connections is created from the logfile.


2. AVAILABILITY
---------------

You can get the latest version of INetSim from
<http://www.inetsim.org>.


3. PREREQUISITES
----------------

- POSIX compatible and System V IPC capable operating system (e.g. Linux)
- Perl version 5.006 or more recent
- Perl library Net::Server
  (available from http://search.cpan.org/~rhandom/Net-Server/)
- Perl library Net::DNS
  (available from http://search.cpan.org/~olaf/Net-DNS/)
- Perl library IPC::Shareable
  (available from http://search.cpan.org/~bsugars/IPC-Shareable/)
- Perl library Digest::SHA
  (available from http://search.cpan.org/~mshelor/Digest-SHA/)
- Perl library IO::Socket::SSL
  (available from http://search.cpan.org/~sullr/IO-Socket-SSL/)
- additionally, for IP-based connection redirection (only supported
  on Linux platforms with Kernel support for packet queueing):
  Perl library Perlipq
  (available from http://search.cpan.org/~jmorris/perlipq/)

INetSim has been developed and tested on Debian GNU/Linux 6.0 and
Ubuntu 12.04 LTS. It has been reported to also run smoothly on
current versions of Gentoo Linux, FreeBSD and OpenBSD.

If you successfully run INetSim on any other platform, or if you
experience problems running INetSim on platforms which meet the
above mentioned requirements, please drop us a note at
<inetsim at inetsim dot org>.


4. INSTALLATION
---------------

Get the latest version of INetSim from
<http://www.inetsim.org>.

Make sure you have installed Perl and all required modules
listed above.

INetSim runs all services with privileges of the user specified
in the configuration file (default: 'nobody'), so make sure
this user exists on your system.
INetSim runs all services with privileges of group 'inetsim',
so you need to add a group with that name to your system.
On a Linux system, this can be done by executing the command
'groupadd inetsim' as root.

Unpack the downloaded tarball to a directory of your choice
(e.g. /opt/inetsim). Change into the top-level directory
of the unpacked tarball and run the script 'setup.sh' as root.
This will set some required permissions on files and
directories of INetSim.


5. CONFIGURATION
----------------

For a documentation of the configuration directives of INetSim,
please refer to the manpage 'inetsim.conf' in subdirectory
'man/man5' and the comments in the sample configuration file
'conf/inetsim.conf' included with the INetSim distribution.


6. USAGE
--------

To start INetSim, change into the top-level directory of the
unpacked tarball and run the startup script 'inetsim' as root.

For a documentation of the available command line options for the
startup script, please refer to the manpage 'inetsim' included
with the INetSim distribution in subdirectory 'man/man1'.

INetSim requires root privileges to bind sockets to ports below 1024.
After binding the sockets, root privileges are dropped as described
in section 'Installation'.
Currently, the INetSim startup script can only be run with root
privileges, even if no ports below 1024 are configured to be used.
This might change in future releases.

!! IMPORTANT NOTE for users of OpenBSD/FreeBSD: !!

The default maximum number of semaphore identifiers on OpenBSD/FreeBSD
is 10. INetSim needs some more semaphore identifiers to run. So you
have to raise the corresponding sysctl value (kern.seminfo.semmni on
OpenBSD, kern.ipc.semmni on FreeBSD). A value of 20 should work.
Otherwise INetSim will crash with an error message like
"Could not create semaphore set: No space left on device" on startup.


7.  ABOUT THE AUTHORS
---------------------

INetSim is developed by Thomas Hungenberg and Matthias Eckert.
We both work in the field of IT security and part of our daily work
is the analysis of unknown malware samples.


8.  ABOUT THE PROJECT
---------------------

To perform a quick run-time analysis of the network behaviour of
unknown malware samples, we were in need of a tool to simulate
internet services which are commonly used by malware in our
laboratory environment. We started off with a bunch of
home-grown Perl scripts together with specially configured
server service implementations like Apache, Postfix, dnsmasq and
ntpd, but we were not happy with this because of a lot of
disadvantages resulting from the combination of many programs
(e.g. problems with correlation of log data).

While talking to other security analysts, we noticed that there
is definitely a need for a comfortable single suite to simulate
different internet services with common logging and centralized
control functions. So we decided to start the project 'INetSim'
to develop such a suite.

Due to lack of time at the office, the programming was done in
our spare time. We both have been using Perl for many years but
mostly for small scripts, e.g. for the analysis of logfiles.
The project INetSim was a welcome opportunity to gain more
practical experience in programming Perl and to deal with the
specifications (RFCs) for several services in depth.

We think INetSim might be useful for other security researchers
as well and therefore decided to release it to the community as
free software licensed under the GNU General Public License (GPL).

Any feedback on your experiences with INetSim is appreciated.
Please send your comments to <inetsim at inetsim dot org>.

NOTE:
As this is our first larger software project written in Perl,
please do not be too harsh when you review the code.
By now, we learned a lot more about using references, packages
and object-oriented programming in Perl, so the design and code
will be much better in our next project. ;-) 


9. COPYRIGHT
------------

Copyright (c) 2007-2014 Thomas Hungenberg & Matthias Eckert

This software is licensed under the GNU General Public License (GPL).
For more information read the file COPYING which should be included
with this distribution.

----------------------------------------------------------------------


CHANGES
1.2.5 (2014-05-24)
------------------
- Fixed incompatibility issue with newer versions of IO::Socket::SSL
- Fixed incompatibility issue with some NTP clients
- Set HTTP(S) POST data directory correctly if '--data-dir' command
  line option is used


1.2.4 (2013-08-15)
------------------
- Fixed broken report module


1.2.3 (2012-10-01)
------------------
- Changed Debian package dependency from Digest::SHA1 to Digest::SHA
- Changed process name format


1.2.2 (2010-11-24)
------------------
- DNS:
  - added check for broken version 0.65 of Net::DNS
- HTTP:
  - added support for static fakefiles
  - some bugfixes with HTTPS fakemode


1.2.1 (2010-07-09)
------------------
- POP3: fixed a small typo that caused the module to be practically unusable


1.2 (2010-04-25)
----------------
- new service module: IRC
  - basic command set
- SMTP:
  - added SSL support (smtps)
  - added support for service extension STARTTLS
  - added support for enforced authentication
- POP3:
  - added SSL support (pop3s)
  - added support for capabilities STLS, TOP, IMPLEMENTATION, LOGIN-DELAY, EXPIRE,
    RESP-CODES and AUTH-RESP-CODE
- HTTP:
  - added SSL support (https)
- FTP:
  - added SSL support (ftps)
- TFTP:
  - added support for TFTP options 'blksize', 'timeout' and 'tsize'
  - added virtual file system
- Redirect:
  - added support for ICMP
- some minor bugfixes


1.1.1 (2009-09-09)
------------------
- FTP: always re-read content of static ftp root directory on startup
- SMTP: set 'Return-Path' to '<>' for MAILER-DAEMON messages,
        replaced header 'Delivered-To' with 'Envelope-To'
- POP3: fixed logging of non-printable characters
- some minor bugfixes


1.1 (2008-10-12)
----------------
- new service module: FTP
- new service module: Syslog
- added support for HTTP methods POST and OPTIONS
- added support for port numbers in HTTP 'Host' header and absoluteURI
- changed SMTP HELO/EHLO response
- added support for SMTP service extensions ENHANCEDSTATUSCODES,
  SEND, SAML, SOML, TURN, ETRN, ATRN, MTRK, BINARYMIME,
  CHUNKING, DELIVERBY, SUBMITTER, CHECKPOINT, NO-SOLICITING, FUTURERELEASE
- added configuration options 'dummy_banner' and 'dummy_banner_wait'
- added configuration options 'redirect_ignore_bootp' and 'redirect_ignore_netbios'
- added configuration options 'create_reports' and 'report_language'
- changed format of service logfile to include service child process id
- changed default value for 'service_timeout' to 120 seconds
- install script 'set_permissions.sh' renamed to 'setup.sh'


1.0 (2008-07-07)
----------------
- changed default setting for configuration option 'redirect_enabled' to 'no'


1.0rc4 (2008-06-29)
-------------------
- Perl library Net::Server is no longer included with the INetSim distribution
- added 'dummy' service module (just logs all received data)
- added 'redirect' feature:
  - uses Linux kernel IP_QUEUE, so only available on Linux platforms
  - allows for IP-based connection redirection (tcp/udp)
  - supports static rules for redirection based on target IP address, port and/or protocol
  - can act as a NAT router for redirection of packets to other hosts
  - optionally varies the TTL value of IP packets sent to the clients from different
    "virtual" connection targets to make traffic look more authentic
- DNS.pm: added handling of AAAA queries (returns empty answer with status 'NOERROR')
- SMTP.pm: add 'Received:' headers to messages
- HTTP.pm: improved Request-URI parsing, now supports absolute URIs
- NTP.pm: added configuration option for less strict checks
- added timeout handling for all service modules
- removed 'bind_address' configuration options for individual services
- some minor bugfixes


1.0rc3 (2007-12-12)
-------------------
- added 'finger' service module
- added configuration option for SMTP and POP3 to select if only
  reversible or all supported authentication mechanisms are offered
- added support for SMTP extensions DSN, ETRN, EXPN, HELP and VERP


1.0rc2 (2007-10-21)
-------------------
- added TFTP service module
- added POP3 configuration options pop3_mbox_maxmails, pop3_mbox_reread
  and pop3_mbox_rebuild
- added sample SPAM mbox file for POP3


1.0rc1 (2007-10-17)
-------------------
- first public release


lib/INetSim.pm
# -*- perl -*-
#
# INetSim - An internet simulation framework
#
# (c)2007-2014 Matthias Eckert, Thomas Hungenberg
#
# Version 1.2.5 (2014-05-24)
#
# For history/changelog see bottom of this file.
#
#############################################################

package INetSim;

use strict;
use warnings;
use POSIX;
# modules to use
use INetSim::CommandLine;
use INetSim::Config;
use INetSim::Log;
use INetSim::FakeTime;
use INetSim::Chargen::TCP;
use INetSim::Chargen::UDP;
use INetSim::Daytime::TCP;
use INetSim::Daytime::UDP;
use INetSim::Discard::TCP;
use INetSim::Discard::UDP;
use INetSim::Echo::TCP;
use INetSim::Echo::UDP;
use INetSim::Quotd::TCP;
use INetSim::Quotd::UDP;
use INetSim::Time::TCP;
use INetSim::Time::UDP;
use INetSim::HTTP;
use INetSim::Ident;
use INetSim::NTP;
use INetSim::SMTP;
use INetSim::POP3;
use INetSim::DNS;
use INetSim::TFTP;
use INetSim::Report;
use INetSim::Finger;
use INetSim::Dummy::TCP;
use INetSim::Dummy::UDP;
use INetSim::FTP;
use INetSim::Syslog;
use INetSim::IRC;

my $VERSION = "INetSim 1.2.5 (2014-05-24)";


#############################################################
# Local variables

my $PPID = $$;   # Parent PID
my @childs = (); # Child PIDs


#############################################################
# Child process handling
#
sub fork_services {
    my @services_to_start = &INetSim::Config::getServicesToStart();
    foreach (@services_to_start) {
	my $pid = fork();
	if ($pid) {
	    # we are the parent process
	    push(@childs, $pid);
	}
	elsif ($pid == 0){
	    # we are the child process
	    if(/^dns$/) {
		&INetSim::DNS::dns;
	    }
	    elsif(/^smtp$/) {
		INetSim::SMTP->run;
	    }
	    elsif(/^smtps$/) {
		INetSim::SMTP->new({ SSL => 1 })->run;
	    }
	    elsif(/^pop3$/) {
		INetSim::POP3->run;
	    }
	    elsif(/^pop3s$/) {
		INetSim::POP3->new({ SSL => 1 })->run;
	    }
	    elsif(/^http$/) {
		INetSim::HTTP->run;
	    }
	    elsif(/^https$/) {
		INetSim::HTTP->new({ SSL => 1 })->run;
	    }
	    elsif(/^ntp$/) {
		INetSim::NTP->run;
	    }
	    elsif(/^time_tcp$/) {
		INetSim::Time::TCP->run;
	    }
	    elsif(/^time_udp$/) {
		INetSim::Time::UDP->run;
	    }
	    elsif(/^daytime_tcp$/) {
		INetSim::Daytime::TCP->run;
	    }
	    elsif(/^daytime_udp$/) {
		INetSim::Daytime::UDP->run;
	    }
	    elsif(/^ident$/) {
		INetSim::Ident->run;
	    }
	    elsif(/^echo_tcp$/) {
		INetSim::Echo::TCP->run;
	    }
	    elsif(/^echo_udp$/) {
		INetSim::Echo::UDP->run;
	    }
	    elsif(/^discard_tcp$/) {
		INetSim::Discard::TCP->run;
	    }
	    elsif(/^discard_udp$/) {
		INetSim::Discard::UDP->run;
	    }
	    elsif(/^chargen_tcp$/) {
		INetSim::Chargen::TCP->run;
	    }
	    elsif(/^chargen_udp$/) {
		INetSim::Chargen::UDP->run;
	    }
	    elsif(/^quotd_tcp$/) {
		INetSim::Quotd::TCP->run;
	    }
	    elsif(/^quotd_udp$/) {
		INetSim::Quotd::UDP->run;
	    }
	    elsif(/^tftp$/) {
		INetSim::TFTP->run;
	    }
	    elsif(/^finger$/) {
		INetSim::Finger->run;
	    }
	    elsif(/^dummy_tcp$/) {
		INetSim::Dummy::TCP->run;
	    }
	    elsif(/^dummy_udp$/) {
		INetSim::Dummy::UDP->run;
	    }
	    elsif(/^ftp$/) {
		INetSim::FTP->run;
	    }
	    elsif(/^ftps$/) {
		INetSim::FTP->new({ SSL => 1 })->run;
	    }
	    elsif(/^syslog$/) {
		INetSim::Syslog->run;
	    }
	    elsif(/^irc$/) {
		INetSim::IRC->run;
	    }
	    elsif(/^ircs$/) {
		INetSim::IRC->run( SSL => 1 );
	    }
	}
	else {
	    &error_exit ("Could not fork: $!", 1);
	}
    }
    sleep 1;
}


sub handle_pid {
    my $cmd = shift;
    my $pidfile = &INetSim::CommandLine::getCommandLineOption("pidfile");

    $pidfile =~ /(.*)/; # evil untaint
    $pidfile = $1;
    if ($cmd eq "create") {
        if (-f $pidfile) {
	    print STDOUT "PIDfile '$pidfile' exists - INetSim already running?\n";
	    exit 1;
	}
	else {
	    if (! open (PIDFILE, "> $pidfile")) {
		print STDOUT "Unable to open PIDfile for writing: $!\n";
		exit 1;
	    }
	    print PIDFILE $PPID;
	    close PIDFILE;
	}
    }
    elsif ($cmd eq "remove") {
        if (-f $pidfile) {
	    unlink $pidfile;
	}
	else {
	    print STDOUT "Hmm, PIDfile '$pidfile' not found (but, who cares?)\n";
	}
    }
}


sub auto_faketime {
    if (&INetSim::Config::getConfigParameter("Faketime_AutoDelay") > 0) {
	my $pid = fork();
	if ($pid) {
	    # we are the parent process
	    push(@childs, $pid);
	}
	elsif ($pid == 0){
	    # we are the child process
	    &INetSim::FakeTime::auto_faketime();
	}
    }
}


sub redirect_packets {
    if (&INetSim::Config::getConfigParameter("Redirect_Enabled")) {
        # check for linux
        if ($^O !~ /linux/i) {
            &INetSim::Log::MainLog("failed! Error: Sorry, the Redirect module does not support this operating system!", "redirect");
            return 0;
        }
        # check for Perlipq library
        eval {
               eval "use IPTables::IPv4::IPQueue; 1" or die;
        };
        if ($@) {
            &INetSim::Log::MainLog("failed! Error: Sorry, this module requires the Perlipq library (IPTables::IPv4::IPQueue)!", "redirect");
            return 0;
        }
        # check for redirect module
        eval {
               eval "use INetSim::Redirect; 1" or die;
        };
        if ($@) {
            &INetSim::Log::MainLog("failed! Error: $@", "redirect");
            return 0;
        }
        my $pid = fork();
        if ($pid) {
	    # we are the parent process
	    push(@childs, $pid);
        }
        elsif ($pid == 0){
	    # we are the child process
	    &INetSim::Redirect::run();
        }
    }
}


sub rest_in_peace {
    my $count = @childs;
    my $i;

    for ($i = 0; $i < $count; $i++) {
        waitpid(-1,&WNOHANG);
        if (! (kill (0, $childs[$i]))) {
            splice (@childs, $i, 1);
            $count = @childs;
            $i--;
        }
    }
}


sub wait_pids {
    wait();
    foreach (@childs){
	waitpid($_, 0);
    }
}


sub kill_pids {
    foreach (@childs){
	kill("TERM", $_);
	waitpid($_, 0);
    }
}


sub error_exit {
    my $msg = shift;
    if (! defined $msg) {
	$msg = "Unknown error";
    }
    my $exitcode = shift;
    if (! defined $exitcode) {
	$exitcode = 1;
    }
    elsif (($exitcode !~ /^[\d]{1,3}$/) || (int($exitcode) < 0) || (int($exitcode > 255))) {
	print STDOUT "Illegal exit code!\n";
	$exitcode = 1;
    }

    print STDOUT "Error: $msg.\n";

    &kill_pids;
    &wait_pids;
    &handle_pid("remove");

    exit 1;
}


#############################################################
# Main
#

sub main {
    # Parse commandline options
    &INetSim::CommandLine::parse_options();

    # Check command line option 'help'
    if (&INetSim::CommandLine::getCommandLineOption("help")) {
	print STDOUT << "EOF";
$VERSION by Matthias Eckert & Thomas Hungenberg

Usage: $0 [options]

Available options:
  --help                         Print this help message.
  --version                      Show version information.
  --config=<filename>            Configuration file to use.
  --log-dir=<directory>          Directory logfiles are written to.
  --data-dir=<directory>         Directory containing service data.
  --report-dir=<directory>       Directory reports are written to.
  --bind-address=<IP address>    Default IP address to bind services to.
                                 Overrides configuration option 'default_bind_address'.
  --max-childs=<num>             Default maximum number of child processes per service.
                                 Overrides configuration option 'default_max_childs'.
  --user=<username>              Default user to run services.
                                 Overrides configuration option 'default_run_as_user'.
  --faketime-init-delta=<secs>   Initial faketime delta (seconds).
                                 Overrides configuration option 'faketime_init_delta'.
  --faketime-auto-delay=<secs>   Delay for auto incrementing faketime (seconds).
                                 Overrides configuration option 'faketime_auto_delay'.
  --faketime-auto-incr=<secs>    Delta for auto incrementing faketime (seconds).
                                 Overrides configuration option 'faketime_auto_increment'.
  --session=<id>                 Session id to use. Defaults to main process id.
  --pidfile=<filename>           Pid file to use. Defaults to '/var/run/inetsim.pid'.

EOF
;
	exit 0;
    }
    elsif (&INetSim::CommandLine::getCommandLineOption("version")) {
        print STDOUT "$VERSION by Matthias Eckert & Thomas Hungenberg\n";
        exit 0;
    }


    # Check if we are running with root privileges (EUID 0)
    if ( $> != 0 ) {
	print STDOUT "Sorry, this program must be started as root!\n";
	exit 1;
    }

    # Check if group "inetsim" exists on system
    my $gid = getgrnam("inetsim");
    if (! defined $gid) {
	print STDOUT "No such group 'inetsim' configured on this system!\n";
	print STDOUT "Please create group and start again. See documentation for more information.\n";
	exit 1;
    }

    print STDOUT "$VERSION by Matthias Eckert & Thomas Hungenberg\n";

    # create pidfile
    &handle_pid("create");

    # Parse configuration file
    &INetSim::Config::parse_config;

    # Check if there are services to start configured, else exit
    if (! scalar(&INetSim::Config::getServicesToStart())) {
	&INetSim::Log::MainLog("No services to start configured. Exiting.");
	&handle_pid("remove");
	exit 0;
    }

    # ignore some signal handlers during startup
    local $SIG{'INT'} = 'IGNORE';
    local $SIG{'HUP'} = 'IGNORE';
    local $SIG{'TERM'} = 'IGNORE';

    &INetSim::Log::MainLog("=== INetSim main process started (PID $PPID) ===");
    &INetSim::Log::MainLog("Session ID:     " . &INetSim::Config::getConfigParameter("SessionID"));
    &INetSim::Log::MainLog("Listening on:   " . &INetSim::Config::getConfigParameter("Default_BindAddress"));
    &INetSim::Log::MainLog("Real Date/Time: " . strftime "%Y-%m-%d %H:%M:%S", localtime);
    &INetSim::Log::MainLog("Fake Date/Time: " . (strftime "%Y-%m-%d %H:%M:%S", localtime(&INetSim::FakeTime::get_faketime())). " (Delta: " . &INetSim::Config::getConfigParameter("Faketime_Delta") . " seconds)");
    &INetSim::Log::MainLog(" Forking services...");
    &fork_services();
    &auto_faketime();
    &redirect_packets();

    if ($$ == $PPID) {
	$0 = 'inetsim_main';
	sleep 2;
        # reap zombies ;-)
        &rest_in_peace;
	&INetSim::Log::MainLog(" done.");
	&INetSim::Log::MainLog("Simulation running.");
	# catch up some signalhandlers for the parent process
	local $SIG{'INT'} = sub {&kill_pids;};
	local $SIG{'HUP'} = sub {&kill_pids;};
	local $SIG{'TERM'} = sub {&kill_pids;};
	&wait_pids;
	&INetSim::Log::MainLog("Simulation stopped.");

	# create report
	if (&INetSim::Config::getConfigParameter("Create_Reports")) {
	    &INetSim::Report::GenReport;
	}

	&INetSim::Log::MainLog("=== INetSim main process stopped (PID $PPID) ===");
	&INetSim::Log::MainLog(".");
    }

    # delete pidfile
    &handle_pid("remove");
    exit 0;
}


1;
#############################################################
#
# History:
#
# Version 1.2.4 (2013-08-14) th
# - changed date/time output format
#
# Version 1.2beta5 (2010-04-15) me
# - added 'Default_BindAddress' to startup screen
#
# Version 1.2beta4 (2009-12-15) th
#
# Version 1.2beta3 (2009-09-23) me  [branch]
# - added service IRC/IRCs
#
# Version 1.1.1 (2009-09-09) th
# - 1.1.1 release
#
# Version 1.2beta2 (2009-09-04) me  [branch]
# - services POP3, HTTP and FTP prepared for using SSL too
#
# Version 1.2beta1 (2009-09-03) me  [branch]
# - added service SMTPS
# - added commandline option '--version' to help output
#
# Version 1.1 (2008-10-12) me
# - 1.1 release
#
# Version 1.1pre4 (2008-09-08) me
# - added service syslog
#
# Version 1.1pre3 (2008-08-24) me
# - changed FTP module name
#
# Version 1.1pre2 (2008-08-20) me
# - added service FTP
#
# Version 1.1pre1 (2008-08-09) th
# - added configuration option 'Create_Reports'
#
# Version 1.0 (2008-07-06) th
# - 1.0 release
#
# Version 1.0rc4 (2008-06-26) th
# - 1.0 rc4 release
#
# Version 1.0rc4pre9 (2008-03-20) me
# - fixed checks in function redirect_packets()
#
# Version 1.0rc4pre8 (2008-03-19) me
# - changed handling for OS check and added an error message
# - removed check for use of 'INetSim::Redirect' (see line below)
# - added check for use of 'IPTables::IPv4::IPQueue' instead of
#   the redirect module and added an error message
# - added function rest_in_peace() to get possible zombies
#
# Version 1.0rc4pre7 (2008-03-17) me
# - added &INetSim::Config::getConfigParameter("Redirect_Enabled") in
#   function redirect_packets()
#
# Version 1.0rc4pre6 (2008-03-15) me
# - added check for use of 'INetSim::Redirect', because it's
#   system dependent
# - added check for operating system in function redirect_packets()
#
# Version 1.0rc4pre5 (2008-03-07) me
# - added service "redirect"
#
# Version 1.0rc4pre4 (2008-03-06) me
# - added service "dummy"
#
# Version 1.0rc4pre3 (2008-03-05) me
# - moved checks for uid '0' and group 'inetsim' below commandline
#   parser, because '--help' should always be possible
#
# Version 1.0rc4pre2 (2008-02-17) me
# - replaced variable 'pidfile' with
#   &INetSim::CommandLine::getCommandLineOption("pidfile");
# - unused old code in function 'handle_pid' removed
#
# Version 1.0rc4pre1 (2007-12-31) th
# - change process names
#
# Version 1.0rc3 (2007-12-12) me
# - new public release
#
# Version 1.0rc3pre1 (2007-11-07) me
# - added service finger
#
# Version 1.0rc2 (2007-10-21) th
# - new public release
#
# Version 1.0rc1 (2007-10-17) th
# - first public release
#
# Version 0.41   (2007-05-31) th
# - added/removed some comments
#
# Version 0.40   (2007-05-23) th
# - added 'sleep 1' to main process to wait for all services started
#
# Version 0.39   (2007-04-30) th
# - check if group 'inetsim' exists on system
# - use getServicesToStart() instead of accessing module
#   variables in Config.pm
#
# Version 0.38   (2007-04-29) th
# - added module CommandLine.pm
# - added check for command line option 'help'
# - added global variable 'VERSION'
#
# Version 0.37   (2007-04-25) th
# - replaced &INetSim::FakeTime::get_faketime_delta() with
#   &INetSim::Config::getConfigParameter("faketime_delta")
#
# Version 0.36   (2007-04-22) th
# - added function error_exit()
# - renamed parameter options in handle_pid() to "create" and "remove"
#
# Version 0.35   (2007-04-21) me
# - added GenReport()
#
# Version 0.34   (2007-04-20) me
# - added check for uid 0
#
# Version 0.33   (2007-04-20) me
# - added function for pidfile handling
#
# Version 0.32   (2007-04-20) me
# - added logging of $INetSim::Config::SessionID
# - re-added a dot at the end of main log
#   (this makes the log easier to read)
#
# Version 0.31   (2007-04-10) th
# - renamed fork_servers() to fork_services()
# - added auto_faketime()
#
# Version 0.30   (2007-04-09) th
# - added module FakeTime
#
# Version 0.30   (2007-04-05) th
# - changed module name 'dns.pm' to 'DNS.pm'
#
# Version 0.29   (2007-03-30) me
# - added service tftp
#
# Version 0.28   (2007-03-27) th
# - moved CONFIGFILENAME, MAINLOGFILENAME and
#   SUBLOGFILENAME to INetSim::Config
# - moved logging functions to INetSim::Log
#
# Version 0.27   (2007-03-26) th
# - changed daytime_tcp to use INetSim::GenericServer
# - changed daytime_udp to use INetSim::GenericServer
# - changed time_tcp to use INetSim::GenericServer
# - changed time_udp to use INetSim::GenericServer
# - changed quotd_tcp to use INetSim::GenericServer
# - changed quotd_udp to use INetSim::GenericServer
# - changed discard_tcp to use INetSim::GenericServer
# - changed discard_udp to use INetSim::GenericServer
# - changed smtp to use INetSim::GenericServer
# - changed pop3 to use INetSim::GenericServer
#
# Version 0.26   (2007-03-24) th
# - changed chargen_tcp to use INetSim::GenericServer
# - changed chargen_udp to use INetSim::GenericServer
# - changed http to use INetSim::GenericServer
#
# Version 0.25   (2007-03-23) th
# - changed echo_tcp to use INetSim::GenericServer
# - changed echo_udp to use INetSim::GenericServer
#
# Version 0.24   (2007-03-19) th
# - added service http
#
# Version 0.23   (2007-03-19) me
# - added service ident
# - added service echo
# - added service discard
# - added service chargen
# - added service discard
#
# Version 0.22   (2007-03-16) th
# - added configuration option @INetSim::Config::ServicesToStart
# - changed BASEDIR to "."
#
# Version 0.21   (2007-03-15) me
# - ignore signals during startup
#
# Version 0.2    (2007-03-15) th
# - added configuration module
# - rewrote fork_servers()
# - changed logfile location
#
# Version 0.1    (2007-03-12) me
#
#############################################################
